scripts/dts/edtlib.py

```diff
@@ -847,111 +822,71 @@ def _init_interrupts(self):
         #
         # , where each phandle points to a node that has a
         #
         #     #<name>-cells = <size>
         #
         # property that gives the number of cells in the value after the
-        # phandle. Also parses any
+        # phandle. These values are given names in #cells in the binding for
+        # the controller.
         #
-        #     <name>-names = "...", "...", ...
-        #
-        # Arguments:
-        #
-        # name:
-        #   The <name>, e.g. "clock" or "pwm". Note: no -s in the argument.
+        # Also parses any
         #
-        # cls:
-        #   A class object. Instances of this class will be created and the
-        #   'node', 'controller', 'data', and 'name' fields initialized. See
-        #   the documentation for e.g. the PWM class.
+        #     <name>-names = "...", "...", ...
         #
-        # Returns a list of 'cls' instances.
+        # Returns a list of ControllerAndData instances.
 
-        prop = self._node.props.get(name + "s")
-        if not prop:
-            return []
+        if prop.name.endswith("gpios"):
+            # There's some slight special-casing for *-gpios properties in that
+            # e.g. foo-gpios still maps to #gpio-cells rather than
+            # #foo-gpio-cells
+            basename = "gpio"
+        else:
+            # Strip -s. We've already checked that the property names end in -s
+            # in _check_binding().
+            basename = prop.name[:-1]
 
         res = []
 
-        for controller_node, spec in _phandle_val_list(prop, name):
-            obj = cls()
-            obj.node = self
-            obj.controller = self.edt._node2enode[controller_node]
-            obj.data = self._named_cells(obj.controller, spec, name)
+        for controller_node, data in _phandle_val_list(prop, basename):
+            mapped_controller, mapped_data = \
+                _map_phandle_array_entry(prop.node, controller_node, data,
+                                         basename)
+
+            entry = ControllerAndData()
+            entry.node = self
+            entry.controller = self.edt._node2enode[mapped_controller]
+            entry.data = self._named_cells(entry.controller, mapped_data,
+                                           basename)
 
-            res.append(obj)
+            res.append(entry)
 
-        _add_names(self._node, name, res)
+        _add_names(self._node, basename, res)
 
         return res
 
-    def _named_cells(self, controller, spec, controller_s):
-        # _init_{interrupts,gpios}() helper. Returns a dictionary that maps
-        # #cell names given in the binding for 'controller' to cell values.
-        # 'spec' is the raw interrupt/GPIO data, and 'controller_s' a string
-        # that gives the context (for error messages).
+    def _named_cells(self, controller, data, controller_s):
+        # Returns a dictionary that maps #cells names given in the binding for
```

Comment:
Now that you are touching this - why not first lookup '#(gpio/interrupt/...)_cells' to get the cells names and if not available use '#cells' as the default fallback. This would also be consistent to the lookup of  '#(gpio/interrupt/...)_cells' to get the number of cells?

Reply:
Yeah, planning to get to it.

Wonder if I should do it in this PR. Getting a bit gargantuan already. :)

Maybe it wouldn't be that big a change though.

Reply:
@b0661 
Think it might be better to do it in a separate PR, because then I could do this:

  1. Convert all bindings to use `interrupt-cells`/`gpio-cells`/etc. instead of `#cells`

  2. Start generating deprecation warnings for `#cells`

Including (1.) in this PR might be a bit much.